department-of-veterans-affairs/vets-api
# Api 29873 confinements date validation

## Summary

- Adds validation for confinement dates:
1. Confinement approximate begin date must be after earliest active duty begin date.
2. Confinement dates must be within one of the service period date ranges.
- FIxes bug in begin_date_is_not_after_end_date?() where we were comparing date strings lexicographically instead of comparing Date objects. 
- Adds rspec tests for validation added.

## Related issue(s)
[API-29873](https://jira.devops.va.gov/browse/API-29873)


## Testing done

- Local testing via Postman:
- Set confinement begin date before earliest active duty begin date & confirm HTTP 422 returned w/ appropriate error message.
- Set confinement begin/end date range outside of any active duty begin/end date range & confirm HTTP 422 returned w/ appropriate error message.

## What areas of the site does it impact?
modules/claims_api/app/controllers/claims_api/v2/veterans/disability_compensation_controller.rb
modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb
modules/claims_api/spec/requests/v2/veterans/disability_compensation_request_spec.rb

## Acceptance criteria
